flow/platform_surface.h

```diff
@@ -0,0 +1,51 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FLOW_PLATFORM_SURFACE_H_
+#define FLUTTER_FLOW_PLATFORM_SURFACE_H_
+
+#include <map>
+#include "flutter/common/threads.h"
+#include "lib/fxl/synchronization/waitable_event.h"
+#include "third_party/skia/include/core/SkMatrix.h"
+#include "third_party/skia/include/core/SkSurface.h"
+#include "third_party/skia/include/gpu/GrBackendSurface.h"
+#include "third_party/skia/include/gpu/GrContext.h"
+#include "third_party/skia/include/gpu/GrTexture.h"
+#include "third_party/skia/include/gpu/GrTypes.h"
+
+namespace flow {
+
+  class PlatformSurface {
+    friend class PlatformSurfaceRegistry;
+   public:
+
+    virtual ~PlatformSurface();
+
+    // Called from GPU thread.
+    virtual sk_sp<SkImage> MakeSkImage(int width,
+                                       int height,
+                                       GrContext* grContext) = 0;
+
+    int Id() { return id_; }
+
+   private:
+    int id_;
+  };
+
+class PlatformSurfaceRegistry {
```

Comment:
I think you can get rid of the `PlatformSurfaceRegistry` completely. You are only using it to hold a reference to the platform handle that you are going to morph into a texture later (since you don't have access to an OpenGL context when the layer tree it built). IMHO, it would be better to reference the platform handle directly (using some sort of the unique_ptr like abstraction that takes a custom deleter) and pass that around instead. You can then convert it to an SkImage when you get to the GPU thread. That way, you won't need the `static`s or the mapping.

Reply:
We tried to do that today, but were not able to follow it through. Maybe we misunderstood you. Our problem seems to be that both the Dart and the Java side need to identify these handles to the engine, the former to tell the engine where in the layer tree the texture should be, the latter to notify the engine that an update (a new video frame) is available and to destroy the handle.

In both cases, we are communicating the identity as an integer. While we could certainly interpret that integer as a C++ heap pointer, we do not see a safe way for the engine to dereference such a pointer without an explicit registry of valid identities. The Dart/Java communication happens asynchronously with layer tree construction, so a Java video plugin could conceivably tear down a handle just as a frame is being rendered based on a layer tree involving that handle.